### PR TITLE
Declare optional dependencies in extras_require

### PR DESCRIPTION
**Pull request by astrofrog**
This does the following:

* Separate testing and non-testing optional dependencies in the install docs
* Declare all optional dependencies in an ``extras_require`` entry in ``setup.py`` - with this, users could get all dependencies by running ``pip install astropy[all]``
* Declare all testing dependencies in the ``test`` entry in ``extras_require``. This would allow all those dependencies to be installed when doing ``pip install astropy[test]``. Note that this is different from ``tests_require`` which is the one actually used to temporarily install the bare minimum testing dependencies when running tests.
* Declare the docs dependencies in the ``docs`` entry - and RTD can recognize this
* Remove the pip requirement files to avoid duplication
* Update the minimum required version of Matplotlib to 2.0 (this is effectively what we supported but the docs claimed 1.5)
* Modify one of the Travis builds to pip install dependencies instead of using ci-helpers

At the moment, this has a small performance impact on Travis - the build I modified took 14 min 37 sec in the last successful run before this PR, and 15 min 58 sec now. I think this is because astropy is built twice, once in the pip install command, and the second time from the ``python setup.py test`` command, so I need to look into whether we can avoid this. Part of it could also be because we now run 12638 tests in this build instead of 12123, presumably because not all optional dependencies were included in ``CONDA_ALL_DEPENDENCIES``.

As a side goal, I want to see how fast installing dependencies with pip is, as if it is quite fast then we could consider actually using pip to install dependencies in most builds instead of conda, and we could also avoid declaring all dependencies in CI config files (instead relying on the minimal/full configurations available in setup.py - that is, some builds could use the ``all`` extras_require`` and some not).

This is an experimental PR that I hope to find time to discuss next week, so do not merge!

### CODE REVIEW

**Comment by drdavella** [comment]
For consistency should `tests_require` contain the same dependencies as `extras_require['test']`? Presumably, `tests_require` is going to be less useful as the community moves away from using `setup.py` as a tool entry point.

**Comment by astrofrog** [comment]
@drdavella - I'm not sure yet, because some of the test dependencies are optional, so I don't want to always force people to temporarily install those. Maybe we should have a ``testall`` extras_require which has all the optional test dependencies, and then ``test`` can match ``tests_require``?

**Comment by drdavella** [comment]
Ah, makes sense. Maybe at the very least we can just say that `extras_require['test'] = tests_require + the_kitchen_sink`.

**Comment by saimn** [comment]
I was thinking about this when playing with tox, but I'm not sure about the 'all' dependencies bundle, which represents a lot of things. It could make sense to have a finer split, maybe by sub-package, 'io', 'table', 'stats', 'viz' ? I think these can be combined in the pip command. But I don't know where to put scipy...

**Comment by bsipocz** [comment]
> But I don't know where to put scipy... 

in that framework I suppose on package can be specified in several points. at least scipy should go into multiple ones

**Comment by saimn** [comment]
> in that framework I suppose on package can be specified in several points. at least scipy should go into multiple ones

That may be an option, but I think if if you specify multiple extra to pip then it will complain about the duplicate dependencies. Something to check.

**Comment by bsipocz** [comment]
Then it's a feature request for them ;)

**Comment by astrofrog** [comment]
It doesn't actually take long to install ``all`` though, so I'm not sure if it's really important to split them up.

**Comment by astrofrog** [comment]
So this now also adds a ``[docs]`` extras require and uses that for RTD. I've also removed pip requirement files because if we do go down this road we should avoid duplicating information.

File: setup.py [hunk 1]
```diff
@@ -84,9 +84,21 @@
 
 install_requires = [min_numpy_version]
 
-extras_require = {
-    'test': ['pytest-astropy']
-}
+extras_require = {}
+
+# All optional dependencies for the tests
+extras_require['test'] = ['pytest-astropy', 'pytest-xdist', 'pytest-mpl',
```
Comment (eteq) [inline setup.py]: We *could* keep the existing requirements files and instead do something like:
```suggestion
with open('pip-requrirements-test') as f:
    extras_require['test'] = [line for line in f.read().split('\n') if line.strip()!='']
```
?

**Comment by pllim** [comment]
> some users are using requirements

Maybe send out a notice to astropy users mailing list to see if there are any objections?

**Comment by astrofrog** [comment]
I've added mentions of this to the installation docs, and tested that it works on RTD. I've also added a dummy pip-requirements file with instructions for using the new way.

File: CHANGES.rst [hunk 2]
```diff
@@ -228,7 +228,11 @@ astropy.wcs
 Other Changes and Additions
 ---------------------------
 
+setup.py
```
[the remark this answers is not in the document]
Comment (astrofrog) [reply CHANGES.rst]: Fixed

**Comment by astrofrog** [comment]
Since the CI passed before my minor changelog entry change, and this has been approved, I'll go ahead and merge.

File: docs/nitpick-exceptions [hunk 3]
```diff
@@ -85,3 +86,8 @@ py:obj RuntimeError
 py:obj NotImplementedError
 py:obj AttributeError
 py:obj NotImplementedError
+py:obj RendererBase
```
Comment (bsipocz) [inline docs/nitpick-exceptions]: I need this change in the release. Should I try to backport the whole PR, or rather just cherry pick this single commit?

**Comment by pllim** [comment]
So, what is the minversion of `setuptools` that supports this?

**Comment by astrofrog** [comment]
I think it's setuptools 30.3.

**Pull request merged**